### ISSUE TITLE
CNTKModel does not output correct result for ResNet50 model when running on Databricks

### ISSUE THREAD

**Issue by memoryz** [Azure/mmlspark#1075]
When we use `CNTKModel` class to load the ResNet50 model and classify an image dataframe, the output of the model is exactly same as the input, as opposed to the logits or probability of the classes. 

This issue only repros for Azure Databricks (5.5 LTS, 6.4 ESR and 8.3), but not for Azure Synapse or local unit tests. This issue also seems to repro only for ResNet50 model (tried CIFAR10, no repro).

**To Reproduce**

```scala
import com.microsoft.ml.spark.image._
import com.microsoft.ml.spark.io.IOImplicits._
import com.microsoft.ml.spark.cntk.CNTKModel
import com.microsoft.ml.spark.downloader.ModelDownloader

val modelDownloader = new ModelDownloader(spark, new java.net.URI("dbfs:/Models/"))
val resnet50 = modelDownloader.downloadByName("ResNet50")

val image_df = spark.read.image.load("wasbs://publicwasb@mmlspark.blob.core.windows.net/explainers/images/dog_cat.jpg")
val cntkModel = new CNTKModel()
  .setModelLocation(resnet50.uri.toString)
  .setInputCol("resize")
  .setOutputCol("out")
  .setOutputNode("outputs")
val requiredSize = cntkModel.getModel.fvar.getArguments.get(0).getShape.getDimensions
val prepare = new ResizeImageTransformer()
  .setInputCol("image")
  .setWidth(requiredSize(0).toInt)
  .setHeight(requiredSize(1).toInt)
  .setNChannels(3)
val resizedDF = prepare.transform(image_df)
val unroll = new UnrollImage()
  .setInputCol(prepare.getOutputCol)
  .setOutputCol("resize")
val unrolledDF = unroll.transform(resizedDF).select("resize").cache
cntkModel.transform(unrolledDF).show
```
**Actual Output**
| resize               | out                  |
|----------------------|----------------------|
| [55.0,48.0,47.0,5... | [55.0,48.0,47.0,5... |

**Expected output**
| resize               | out                  |
|----------------------|----------------------|
| [55.0,48.0,47.0,5... | [1.76913648175514... |

**Info:**
 - MMLSpark Version: 1.0.0-RC3
 - Spark Version 2.4.5 - 3.1.1
 - Spark Platform: **Azure Databricks only**

**Additional context:**

@mhamilton723 traced the problem down to `FlattenBatch` class, and discovered that if the input dataframe is cached, the issue goes away. I'm submitting a PR to apply this workaround, while documenting the issue here for reference and tracking.

**Comment by memoryz** [comment]
Update: 
We reduced the repro to the following lines of code:
```scala
import org.apache.spark.sql.catalyst.encoders.RowEncoder
val df1 = Seq((1,2,3,4)).toDF("col1", "col2", "col3", "col4")
val df2 = df1.mapPartitions(identity)(RowEncoder(df1.schema))
val df3 = df2.drop("col2")
val df4 = df3.mapPartitions(identity)(RowEncoder(df3.schema))
display(df4)
```

Expected result:
| col1 | col3 | col4 |
|------|------|------|
|1|3|4|

Actual result:
| col1 | col3 | col4 |
|------|------|------|
|1|2|3|

Environment: Azure Databricks (does not reproduce on Azure Synapse).

Databricks runtime version: 5.5 LTS, 6.4 ESR, 7.3 LTS, 8.2, 8.3, 8.4
